Add interfaces for extension managements & translation hooks

haq204 opened the pull request:
The Manager interface handles and maintains the registered extensions. It contains methods to determine whether an API group is supported by a registered extension and methods to return hook clients for different hook types (e.g. XDSHook). 
The Manager is intended to be created upon runner startup, load the registered extensions from the `EnvoyConfig` which the runner has access to, and pass it as an input the main Translate function. The Translate function can then use the manager when translating to determine whether it should mark an API group found in the extensionRef whether a particular extension will be handling it.

IR
```go
import (
    "sigs.k8s.io/gateway-api/apis/v1beta1"
)

type HTTPRoute struct {
    Name string
    ...

    FilterExtensionRefs []ExtensionRef
}

type ExtensionRef struct {
    ExtensionName string
    Resource *v1beta1.LocalObjectReference
}
```

Translate
```golang
import (
    "github.com/envoyproxy/gateway/internal/ir"
    xdsTypes "github.com/envoyproxy/gateway/internal/xds/types"
    extTypes "github.com/envoyproxy/gateway/internal/extension/types"
)

// Translate translates the XDS IR into xDS resources
func Translate(ir *ir.Xds, em extTypes.Manager) (*xdsTypes.ResourceVersionTable, error) {
    if ir == nil {
        return nil, errors.New("ir is nil")
    }

    tCtx := new(types.ResourceVersionTable)
    httpExtRefs := new(extTypes.HTTPListenerExtensionRefs)

    // httpExtRefs would be updated for HTTPRouteFilter extensionRefs to known extension resources.
    if err := processHTTPListenerXdsTranslation(tCtx, ir.HTTP, httpExtRefs); err != nil {
        return nil, err
    }

    if err := processPostHTTPListenerTranslationHooks(tCtx, httpExtRefs, em); err != nil {
        return nil, err
    }

    if err := processTCPListenerXdsTranslation(tCtx, ir.TCP); err != nil {
        return nil, err
    }

    if err := processUDPListenerXdsTranslation(tCtx, ir.UDP); err != nil {
	return nil, err
    }
    return tCtx, nil
}
```

Each type of translation hook has its own interface that contains methods representing different types of hook entrypoints for that particular translation step. The hook clients returned by the `Manager` abstracts the specific protocol we are using to communicate to the extension service (gRPC).

Finally, we add some of the proto objects that the Envoy Gateway Control Plane and the extension service will be interacting with. These objects provide the context needed for the extension to modify the resource according to user intent. The actual gRPC service spec will be specified in a subsequent PR.

Signed-off-by: Hamzah Qudsi <hqudsi@datawire.io>

LanceEa commented on internal/extension/proto/extension.proto:
```diff
@@ -0,0 +1,56 @@
+message HTTPRouteFilterExtensionContext {
+    // route_name is the name of HTTPRoute the filter is being applied on
+    //
+    // TODO: Will we need more context than just route name such as Path, Headers,
```
On your comment: Lets explore trying to do this just with RouteName. Note: this does mean the Gateway project will need to ensure routes are named properly and unique which you will see that currently not all cases give a route a unique name

### Why? 
The reason for having EG as the foundation is so that everyone that wants to extend it doesn't have to duplicate work/algorithms. EG already handles the route matching so we just need to provide the Hook enough information to know which route the extensionRef is referring to.

If we find uses cases where that falls down or is sufficient then we can explore but for a v0 I want us to try and explore ways where we are not duplicating EG logic.

haq204 replied:
yea think just matching by HTTPRoute name and hostname should be sufficient for now.

I guess it's going to depend how `HTTPRoutes` are named in the IR. If it's by `namespace/HTTPRouteName` that should be enough to guarantee uniqueness because K8s doesn't allow you to have the same name for different objects of the same Kind within the same namespace. which looks like it does https://github.com/datawire/envoy-gateway/blob/8dd60b1a3aaa8a93504aae0efd4d50315a43bd96/internal/gatewayapi/helpers.go#L365.

This is appended in the ir.HTTPRoute.Name along with the host. https://github.com/datawire/envoy-gateway/blob/8dd60b1a3aaa8a93504aae0efd4d50315a43bd96/internal/gatewayapi/route.go#L260

LanceEa replied:
Yes, in the IR but it doesn't flow through to every Envoy Route that is created as seen here:

https://github.dev/envoyproxy/gateway/blob/c007313d7ec81d7728b847b62f1b513db93e4d73/internal/xds/translator/route.go#L38 

So, I think it works but we would have to modify it so the routes added to the VHost included the `.Name`. Now that probably gets tricky with mirroring and stuff now that I think about it....

haq204 replied:
we've decided that `route_name` and `hostnames` are sufficient for the time being

LanceEa commented on internal/extension/proto/extension.proto:
```diff
@@ -0,0 +1,56 @@
+    string route_name = 1;
+
+    // hostnames are the fully qualified domain names attached to the HTTPRoute
+    repeated string hostnames = 2;
```
The reason this is here is so that we can quickly zero in on the correct FilterChain that this rule applies...this to me is in the same category as the route_name comment above. Maybe a better approach is to give the FilterChains unique names and pass that instead. Open to other ideas.

haq204 replied:
Enforcing uniqueness on filter chain names could work. I think that would be  easiest vs. sending over hostname, and possibly other match types like path, header, query params and make the extension service figure out which filter chain it should apply to. Envoy Gateway fully controls the names of filter chains and doesn't expose that anywhere else (i.e it doesn't surface in the xdsIR).

LanceEa commented on internal/extension/proto/extension.proto:
```diff
@@ -0,0 +1,56 @@
+
+    // route_filter_extension_refs contains the list of Objects added as extensionRefs
+    // to an HTTPRouteFilter.
+    repeated ObjectReference route_filter_extension_refs = 3;
```
Just for clarity sake, the reason this is repeated is because AFAIK Gateway API (HTTPRoute) doesn't prohibit apply many httpFilters of type ExtensionRef

haq204 replied:
correct

LanceEa commented on internal/extension/proto/extension.proto:
```diff
@@ -0,0 +1,56 @@
+
+// InfraManagerContext provides context on the Gateway deployment
+// so that it can be modified by the Extension.
+message InfraManagerContext {
```
Let's just drop this for now. Its not in scope for v0. Then add it when we are ready. We can capture it in a design doc so that its clear we are thinking about it.

LanceEa commented on internal/extension/types/tables.go:
```diff
@@ -0,0 +1,15 @@
+)
+
+// HTTPListenerExtensionRefs holds all HTTPRouteFilter extensionRefs along with
+// its corresponding context from the associated HTTPRoute. 
```
Can you add to the comment what the Key represents so its clear?

LanceEa commented on internal/extension/types/manager.go:
```diff
@@ -0,0 +1,47 @@
+	// GetExtension returns an Extension for the provided extension id.
+	//
+	// If no Extension exists for that id, then it should return (nil, error).
+	GetExtension(eid string) (*Extension, error)
```
Is this being added for completeness or is there a specific purposed for exposing it?

haq204 replied:
Mostly for completeness but an implementation could be making use of this. For example to implement `GetXDSHookClient` it might call `m.GetExtension(eid)` to get the extension service to instantiate the transport client. And I see no reason why that has to be private.

I also consider possibly separating the extension management and hook management methods into separate interfaces. so when we do eventually add the InfraManager hook it won't immediately break tests because the Manager interface changed

```golang
type ExtensionManager interface {
       HasExtension(g v1beta1.Group, k v1beta1.Kind) (bool *string)

       GetExtension(eid string) (*Extension, error)

       ListExtensions() []*Extension
}

type XDSHookManager interface {
     GetXDSHookClient(eid string) (*XDSHookClient, error)
}
```

In the beginning we can omit `Get` and `List` methods and add it back only if we find it useful. Implementations can add them if they want as helper methods

LanceEa commented on internal/extension/types/manager.go:
```diff
@@ -0,0 +1,47 @@
+	GetExtension(eid string) (*Extension, error)
+
+	// ListExtensions returns a list of all extensions registered by the Manager.
+	ListExtensions() []*Extension
```
Same on this one, The only use case I can think of for exposing this is for logging or is there any reason to have this?

haq204 replied:
see above comment

LanceEa commented on internal/extension/types/hooks.go:
```diff
@@ -0,0 +1,35 @@
+// the InfraManager translation pipeline.
+//
+// WARNING: This interface isn't finalized at all and subject to change
+type InfraManagerHookClient interface {
```
Let's drop Infra stuff at this time. Yes, we can explore it once we have a v0 but I would rather it not distract from getting a working v0

LanceEa commented on internal/extension/types/hooks.go:
```diff
@@ -0,0 +1,35 @@
+type XDSHookClient interface {
+	// PostHTTPListenerTranslation is the entrypoint for modifying XDS after processing
+	// the HTTP listener.
+	PostHTTPListenerTranslation(eCtx *proto.HTTPListenerExtensionContext, tCtx *xdsTypes.ResourceVersionTable) error
```
Since you are not returning the ResouceVersionTable,  is the expectation that the implementation of the client will modify the data for the pointer to that? If so then isn't the implementation going to be difficult because you would have to deal with some kind of merge semantics from what the extension sent back or you would just swap it completely? 

I guess what I'm driving at is that reading this function it isn't clear to me what the expected outcome would be?

haq204 replied:
see below comment

LanceEa commented on internal/extension/types/hooks.go:
```diff
@@ -0,0 +1,35 @@
+type XDSHookClient interface {
+	// PostHTTPListenerTranslation is the entrypoint for modifying XDS after processing
+	// the HTTP listener.
+	PostHTTPListenerTranslation(eCtx *proto.HTTPListenerExtensionContext, tCtx *xdsTypes.ResourceVersionTable) error
```
Sort of a bikeshed: personally not a fan of these names because it requires additional mental load to figure out what they are referring to. `eCtx` makes a little bit of sense but the `tCtx` doesn't because its referring to untyped envoy config sources.

Second, Is the intent here to send the whole ResourceVersionTable for every listener? This would mean we would be serializing and sending over the wire a bunch of routes/listeners/clusters/ secrets that have nothing to do with it. Which made me just think...sending the ResourceVersionTable would also include secrets which I'm not so certain we want to be sending secrets over to an extension...

haq204 replied:
tbh I'm not satisfied with the names either so if you have better names, I'm open to suggestions. `tCtx` is what is generally been used across the project to refer to the Envoy resource table

On the second point, remember that there are two interfaces `XDSHookClient` and the gRPC client interface to the extension service, call it `ExtensionXDSClient`.

`XDSHookClient` is the internal data structure that will be actually used by the translators, abtracting the gRPC transport - so its implementation will utilize `ExtensionXDSClient`. `ExtensionXDSClient` might have its own `PostHTTPListenerTranslation` method but with gRPC specific arguments

```go
type ExtensionXDSClient interface {
    PostHTTPListenerTranslation(ctx context.Context, in *PostHTTPListenerTranslationRequest, opts ...grpc.CallOption) (*PostHTTPListenerTranslationResponse, error)
}
```

`PostHTTPListenerTranslationRequest` would wrap `HTTPListenerExtensionContext`. Notice that we didn't send `ResourceVersionTable` over the wire. The extension would return `PostHTTPListenerTranslationResponse` that may outline some XDS-like data structure or even wrap its own `ResourceVersionTable` and the implementation of `XDSHookClient.PostHTTPListenerTranslation` would handle merging it with `ResourceVersionTable` after it receives the gRPC response.

Of course we can also do what you suggested and have the merge occur outside the method:
```go
tCtx := make(ResourceVersionTable)
// some stuff occurs
resp := xdsClient.PostHTTPListenerTranslation(eCtx)
// merge resp with tCtx
```

haq204 replied:
update godoc

LanceEa reviewed (commented):
Look good, I have let some general comments and then also needing some clarity to understand the plan for a few of the items. Lets have @AliceProxy , you and me jump on a call when she is in to just hash through these.

Commit:
build: simplify buf workspace

Simplify buf workspace into a single module

Signed-off-by: Hamzah Qudsi <hqudsi@datawire.io>

Commit:
feature: add interfaces for extension managements & translation hooks

The Manager interface handles and maintains the registered extensions. It contains methods to determine whether an API group is supported by a registered extension and methods to return hook clients for different hook types (e.g. XDSHook)

Each type of translation hook has its own interface that contains methods representing different types of hook entrypoints for that particular translation step.

Finally, we some of the proto objects that the Envoy Gateway Control Plane and the extension service will be interacting with. These objects provide the context needed for the extension to modify the resource according to user intent. The actual gRPC service spec will be specified separately.

Signed-off-by: Hamzah Qudsi <hqudsi@datawire.io>

Commit:
feature: add gRPC API for extension service

Signed-off-by: Hamzah Qudsi <hqudsi@datawire.io>

Alice-Lilith reviewed (approved):
lgtm
